src/arizona_template_compiler.erl

```diff
@@ -0,0 +1,653 @@
+compile_elems([{expr, Loc, Expr} | T], State) ->
+    compile_expr(Expr, Loc, T, State);
+compile_elems([{block, Loc, Block} | T], State) ->
+    compile_block(Block, Loc, T, State);
+compile_elems([{tag, Loc, #{is_stateful := true}} | _T], State) ->
+    throw({unexpected_stateful_tag, Loc, State});
+compile_elems([{tag, Loc, Tag} | T], State) ->
+    compile_tag(Tag, Loc, T, State);
+compile_elems([], _State) ->
+    [].
+
+compile_text(Txt, T, State) ->
+    [{text, Txt} | compile_elems(T, State)].
+
+compile_expr(ExprStr, Loc, T, State) ->
+    case expand_expr(ExprStr, Loc, true, State) of
+        {text, Txt} ->
+            [{text, Txt} | compile_elems(T, State)];
+        {expr, Expr} ->
+            [{expr, Expr} | compile_elems(T, incr_index(1, State))]
+    end.
+
+expand_expr(Expr, Loc, Eval, State) ->
+    ExprTree = merl:quote(Expr),
+    AttrsEnv = lists:map(fun
+        ({K, {expr, _KLoc, KExpr}}) ->
+            {binary_to_existing_atom(K, utf8), merl:quote(KExpr)};
+        ({K, {text, _KLoc, KTxt}}) ->
+            {binary_to_existing_atom(K, utf8),
+              merl:quote(iolist_to_binary(["<<\"", KTxt, "\"/utf8>>"]))}
+    end, State#state.attributes),
+    AttrsTree = merl:tsubst(ExprTree, AttrsEnv),
+    Macros = State#state.macros,
+    MacrosEnv = [{K, merl:term(V)} || K := V <- Macros],
+    MacrosTree = merl:tsubst(AttrsTree, MacrosEnv),
+    AllVars = merl:template_vars(merl:template(MacrosTree)),
+    case AllVars -- maps:keys(Macros) of
+        [] when Eval ->
+            try
+                Form = erl_syntax:revert(merl:tree(MacrosTree)),
+                {value, Value, _} = erl_eval:exprs([Form], []),
+                {text, arizona_html:to_safe(Value)}
+            catch
+                _:_ ->
+                    throw({invalid_macro_value, Loc, State})
+            end;
+        Vars ->
+            VarsSubst = [{Var, var_form(Var)} || Var <- Vars],
+            Env = [{subst, merl:subst(MacrosTree, VarsSubst)}],
+            [Form] = erl_syntax:revert_forms([merl:qquote(~"_@subst", Env)]),
+            NormExpr = iolist_to_binary(erl_pp:expr(Form)),
+            Forms = [merl:quote(<<"fun(Assigns) -> ", NormExpr/binary, " end">>)],
+            {value, Fun, _} = erl_eval:exprs(Forms, []),
+            {expr, expr(Fun, Vars, State)}
+    end.
+
+var_form(Var) ->
+    VarName = atom_to_binary(Var, utf8),
+    merl:quote(<<"maps:get(", VarName/binary, ", Assigns)">>).
+
+expr(Fun, Vars, State) ->
+    #{
+        id => changeable_id(State),
+        function => Fun,
+        vars => Vars
+     }.
+
+compile_block(#{name := Name} = Block, Loc, T, State) ->
+    case block_mod_fun(Name, State) of
+        {ok, {Mod, Fun}} ->
+            Attrs = maps:get(attributes, Block),
+            Elems = expand_block(Mod, Fun, Attrs, State),
+            Elems ++ compile_elems(T, incr_index(changeable_count(Elems, 0), State));
+        {error, Reason} ->
+            throw({Reason, Loc, State})
```

Comment:
Would it be too much trouble to have the `throw` in the same context? I think I understand what you're trying to do, but having these jump all around is not easy to follow (or potentially maintain, in the future).

Reply:
[c3c85f9](https://github.com/arizona-framework/arizona/pull/164/commits/c3c85f948bd7f2b6fe28802b65ebf90c4517c5cb)